src/lib.rs

```diff
@@ -699,4 +824,99 @@ mod unix {
             .try_into()
             .context("Failed to get page size")
     }
+
+    impl PageFaultChecker {
+        pub(super) fn new(ptr: *mut usize, len: usize) -> PageFaultChecker {
+            PageFaultChecker {
+                ptr,
+                len,
+                state: None,
+            }
+        }
+    }
+
+    unsafe impl Send for PageFaultChecker {}
+
+    impl super::memtest::TestObserver for PageFaultChecker {
+        type Error = PageFaultError;
+
+        // TODO: Currently NUM_CHECKS is an arbitrary number
+        fn init(&mut self, expected_iter: u64) {
+            const FIRST_CHECKPOINT: u64 = 0;
+            const NUM_CHECKS: u64 = 100;
+
+            assert!(
+                self.state.is_none(),
+                "init() should only be called once per test"
+            );
+
+            self.state = Some(PageFaultCheckerState {
+                completed_iter: 0,
+                checkpoint: FIRST_CHECKPOINT,
+                checkpoint_step: expected_iter / NUM_CHECKS,
+            });
+        }
+
+        #[inline(always)]
+        fn check(&mut self) -> Result<(), Self::Error> {
+            let state = self
+                .state
+                .as_mut()
+                .expect("init() should be called before check()");
+
+            if state.completed_iter < state.checkpoint {
+                state.completed_iter += 1;
+                return Ok(());
+            }
+
+            state.on_checkpoint(self.ptr, self.len)
+        }
+    }
+
+    impl PageFaultCheckerState {
+        fn on_checkpoint(&mut self, ptr: *mut usize, len: usize) -> Result<(), PageFaultError> {
+            use libc::mincore;
+
+            unsafe {
+                let page_size = get_page_size().unwrap();
+                let base_address = ptr as usize;
+                let mem_size = len * size_of::<usize>();
+
+                let aligned_address = base_address & !(page_size - 1);
+                let aligned_ptr = (aligned_address as *mut usize).cast();
+                let aligned_mem_size = mem_size + (base_address - aligned_address);
+
+                // buf needs to be at least (aligned_mem_size + page_size - 1) / page_size
+                let mut buf: Vec<u8> = vec![0; aligned_mem_size.div_ceil(page_size)];
+                assert_eq!(
+                    mincore(aligned_ptr, aligned_mem_size, buf.as_mut_ptr().cast()),
+                    0,
+                    "mincore did not return 0"
+                );
+                if buf.iter().any(|n| *n == 0) {
```

Comment:
Uh oh -- This is a logic error. According to the documentation for `mincore()`, only the least significant bit of each of these bytes is set or cleared -- The rest of them are undefined.